README

This is an implementation of the simple, but fun board game known as
Mancala.  For rules of the game, see RULES.

A simple character/terminal-based version of the game can be compiled
on both Un*x- and MS-DOS-systems.  A fancy windowing version requires
the X Window System.  See INSTALL for brief instructions on how to
build the programs.

This package is brought to you by `Fellowship of Beer-drinking
Programmers', currently

	Sverre H.  Huseby	shh@thathost.com
	Glenn Terje Lines	glennli@simula.no

Feel free to do what you want with this software, but please drop any
of us a mail if you use it (or if you don't use it, for some reason).
If you make any changes to the code before passing it further, you
must include a file identifying what changes you made.  Please leave
our names where you find'em.  Fame and fortune, you know...


NEWS
mancala 1.0.3
-------------

Crash fixes.


mancala 1.0.2
-------------

Major changes in this version (there really aren't any, so here comes
the minor changes):

  * Updated to work with more recent version of XForms.

  * Added a man page.

  * Added NEWS (this file!) and ChangeLog.

  * Builds on OS X with MacPorts.

Most changes were provided by the Debian maintainer Paul Wise.


mancala.c
/* $Id: mancala.c,v 1.2 2007-06-13 18:29:52 sverrehu Exp $ */
/**************************************************************************
 *
 *  FILE            mancala.c
 *  MODULE OF       The board game Mancala.
 *
 *  DESCRIPTION     An implementation of the simple game called Mancala.
 *                  This is the backend.
 *
 *  WRITTEN BY      Sverre H. Huseby
 *
 **************************************************************************/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <time.h>

#include "minimax.h"
#include "mancala.h"

/**************************************************************************
 *                                                                        *
 *                       P R I V A T E    D A T A                         *
 *                                                                        *
 **************************************************************************/

/*
 *  Define this if you want a more timeconsuming evaluation function.
 *  We're not sure it makes it any better -- it may even make it worse.
 */
#define FULL_EVAL

/*
 *  The board state stack, and the current stacklevel.
 */
static Board boardStack[STACK_LEVELS];
static int   idx = 0;

/*
 *  The game logic function requests an array of pointers to possible
 *  moves for a level of it's recursion. We need one array for each
 *  possible recursion level, so we don't mess things up. possibleMoveTable
 *  contains the real move variables, while possibleMove contains the
 *  pointers into the variable table.
 */
static Move  possibleMoveTable[STACK_LEVELS][MAX_HOLES];
static PMove possibleMove[STACK_LEVELS][MAX_HOLES];



/**************************************************************************
 *                                                                        *
 *                   P R I V A T E    F U N C T I O N S                   *
 *                                                                        *
 **************************************************************************/

/*------------------------------------------------------------------------*
 |                 Callbacks for the game logic function                  |
 *------------------------------------------------------------------------*/

/*-------------------------------------------------------------------------
 *
 *  NAME          pushBoard
 *
 *  FUNCTION      Save the current board state on a stack.
 *
 *  DESCRIPTION   This callback is called at the start of each recursion
 *                level. We use it to save the current board setup, and
 *                to make sure requests for new moves (getMoves()) don't
 *                mess up the array of moves currently being explored.
 */
static void pushBoard(void)
{
    ++idx;
    memcpy(&boardStack[idx], &boardStack[idx - 1], sizeof(Board));
}



/*-------------------------------------------------------------------------
 *
 *  NAME          popBoard
 *
 *  FUNCTION      Restore the current board stat from the stack.
 *
 *  DESCRIPTION   This callback is called at the end of each recursion
 *                level. We restore the board state (remove the
 *                `experimental' moves), and open for reuse of the array
 *                of possible moves.
 */
static void popBoard(void)
{
    --idx;
}



/*-------------------------------------------------------------------------
 *
 *  NAME          getMoves
 *
 *  FUNCTION      Make list of possible moves for the given player.
 *
 *  INPUT         player  the player to make movelist for.
 *
 *  OUTPUT        numMoves
 *                        number of moves in the returned array.
 *
 *  RETURNS       An array of pointers to legal moves, or NULL if no
 *                legal moves for this player.
 *
 *  DESCRIPTION   This callback is called for each recursion level to
 *                get a list of possible following moves. A legal move is
 *                the number of a non-empty hole for the player in question.
 *
 *                The current stacklevel is used to find a `free' array
 *                that we can fill with legal moves. We couldn't use the
 *                same array all the time, since the array we returned in
 *                the previous call is not fully examined (yeah, recursion
 *                gives some interresting problems...).
 */
static PMove *getMoves(Player player, int *numMoves)
{
    int        q, n = 0;
    StoneCount *hole;
    PMove      *m, *ret;

    m = ret = possibleMove[idx];
    hole = boardStack[idx].hole[player];
    for (q = 0; q < MAX_HOLES; q++) {
	if (*hole)
	    m[n++]->hole = q;
	++hole;
    }
    if ((*numMoves = n) == 0)
	ret = NULL;
    return ret;
}



/*-------------------------------------------------------------------------
 *
 *  NAME          undoMove
 *
 *  FUNCTION      Update the board to the state before the given move.
 *
 *  INPUT         player  the player that did the move.
 *                move    pointer to the move to undo.
 *
 *  DESCRIPTION   This callback is called at the end of each recursion
 *                level. Since we really change the board in popBoard(),
 *                we don't need to do anything here.
 */
static void undoMove(Player player, PMove move)
{
    /* memcpy(&boardStack[idx], &boardStack[idx - 1], sizeof(Board)); */
}



/*-------------------------------------------------------------------------
 *
 *  NAME          evalBoard
 *
 *  FUNCTION      Evaluate the board after a player has moved.
 *
 *  INPUT         player  the player to evaluate the board for.
 *
 *  RETURNS       A measure indicating how good the current board is for
 *                the given player. Larger values are better. A positive
 *                value indicates that the given player is in the lead,
 *                while a negative value indicates the opposite.
 *
 *  DESCRIPTION   We do it simple; just see which player has captured
 *                most stones.
 */
static Score evalBoard(Player player)
{
    Board  *b;

#ifdef FULL_EVAL
    Player winner;
    /*
     *  The next thing that happens after the evaluation, is that the
     *  previous board is popped from the stack, so we can let the
     *  evaluation function mess the board up like this:
     */
    checkAndFixWin(&winner);
#endif
    b = &boardStack[idx];
    return b->mancala[player] - b->mancala[player ^ 1];
}



/**************************************************************************
 *                                                                        *
 *                    P U B L I C    F U N C T I O N S                    *
 *                                                                        *
 **************************************************************************/

/*-------------------------------------------------------------------------
 *
 *  NAME          initGame
 *
 *  FUNCTION      Set up variables for a new game.
 */
void initGame(int stones_pr_hole)
{
    int    q, w;
    Board  *b;
    Player player;
    
    /*
     *  The game logic function uses random numbers to choose between
     *  moves that gives equal score, so we initiate the generator.
     */
    srand(time(NULL));

    /*
     *  An important step: Set up the possible move -pointers that the
     *  caller of getMoves() want. We make pointers into the table
     *  containing the real values.
     */
    for (q = 0; q < STACK_LEVELS; q++)
	for (w = 0; w < MAX_HOLES; w++)
	    possibleMove[q][w] = &possibleMoveTable[q][w];

    /*
     *  Clear the game stack, and set up the board to initial state.
     *  The `real' board is actually at position 0 of the stack array.
     */
    idx = 0;
    b = &boardStack[idx];
    for (player = 0; player < 2; player++) {
	b->mancala[player] = 0;
	for (w = 0; w < MAX_HOLES; w++)
	    b->hole[player][w] = stones_pr_hole;
    }
}



/*-------------------------------------------------------------------------
 *
 *  NAME          getHole
 *
 *  FUNCTION      Get number of stones in a hole on the board.
 *
 *  INPUT         player  the player to get number of stones for.
 *                hole    the hole number. 0 is the hole closest to
 *                        the mancala.
 *
 *  RETURNS       The number of stones in the given hole.
 */
StoneCount getHole(Player player, int hole)
{
    return boardStack[0].hole[player][hole];
}



/*-------------------------------------------------------------------------
 *
 *  NAME          getMancala
 *
 *  FUNCTION      Get number of stones in the mancala of a player.
 *
 *  INPUT         player  the player to get number of stones for.
 *
 *  RETURNS       The number of stones in the given player's mancala.
 */
StoneCount getMancala(Player player)
{
    return boardStack[0].mancala[player];
}



/*-------------------------------------------------------------------------
 *
 *  NAME          legalMove
 *
 *  FUNCTION      Check if a human's move is legal.
 *
 *  INPUT         player  the player that wants to move.
 *                move    pointer to the move to check.
 *
 *  RETURNS       1 if legal move, 0 if illegal.
 *
 *  DESCRIPTION   This is used to check if a human player has chosen to
 *                do a legal move. The computer player will only do moves
 *                returned by getMoves(), and those are (or should) all be
 *                legal.
 */
int legalMove(Player player, PMove move)
{
    int   q;
    PMove *legalmove;     /* array of pointerts to legal moves */
    int   numMoves;       /* number of moves in this array */

    legalmove = getMoves(player, &numMoves);
    for (q = 0; q < numMoves; q++) {
	if ((*legalmove)->hole == move->hole)
	    return 1;
	++legalmove;
    }
    return 0;
}



/*-------------------------------------------------------------------------
 *
 *  NAME          checkAndFixWin
 *
 *  FUNCTION      Check if the game is ended, and in that case fix scores.
 *
 *  OUTPUT        winner  the player that won.
 *
 *  RETURNS       0 if game not finished, 1 if one winner, 2 if draw.
 *
 *  DESCRIPTION   The game is over if any of the players have no stones
 *                left to move. In that case, all remaining stones for the
 *                opposite player is moved to that player's mancala, and
 *                the winner is the one with more stones.
 */
int checkAndFixWin(Player *winner)
{
    int        q, ret = 0, sum[2];
    Board      *b;
    StoneCount *hole;
    Player     player;

    b = &boardStack[idx];
    for (player = 0; player < 2; player++) {
	hole = b->hole[player];
	sum[player] = 0;
	for (q = 0; q < MAX_HOLES; q++)
	    sum[player] += hole[q];
    }
    if (sum[0] == 0 || sum[1] == 0) {
	ret = 1;
	for (player = 0; player < 2; player++) {
	    b->mancala[player] += sum[player];
	    hole = b->hole[player];
	    for (q = 0; q < MAX_HOLES; q++)
		hole[q] = 0;
	}
	if (b->mancala[0] > b->mancala[1])
	    *winner = 0;
	else if (b->mancala[0] < b->mancala[1])
	    *winner = 1;
	else
	    ret = 2;
    }
    return ret;
}



/*-------------------------------------------------------------------------
 *
 *  NAME          getBestMove
 *
 *  FUNCTION      Calculate best move for player.
 *
 *  INPUT         player  the player to evaluate.
 *                maxPly  max recursion depth.
 *
 *  RETURNS       Pointer to the best move to make, or NULL if no move
 *                available.
 *
 *  DESCRIPTION   This is just a frontend to the real game logic function,
 *                hiding the static functions found in this file.
 */
PMove getBestMove(Player player, int maxPly)
{
    return miniMax(pushBoard, popBoard,
		   getMoves, doMove, undoMove,
		   evalBoard, player, maxPly);
}



/*------------------------------------------------------------------------*
 |   The following function is a callback, but it's used for moving too.  |
 *------------------------------------------------------------------------*/

/*-------------------------------------------------------------------------
 *
 *  NAME          doMove
 *
 *  FUNCTION      Update the board with the given move.
 *
 *  INPUT         player  the player that does the move.
 *                move    pointer to the move to make.
 *
 *  RETURNS       The next player to move. For Mancala, this is the same
 *                player once again if the last stone ends in the mancala.
 *
 *  DESCRIPTION   This function is used both as a callback for finding the
 *                best moves, and as a `normal' function to update the
 *                board when a move is chosen.
 */
Player doMove(Player player, PMove move)
{
    int        seeds, holeNum, oppositeHoleNum;
    StoneCount *hole, *oppositeHole;
    Board      *b;
    Player     nextPlayer, holeOwner;

    nextPlayer = player ^ 1;
    b = &boardStack[idx];
    holeOwner = player;
    hole = b->hole[holeOwner];
    holeNum = move->hole;
    seeds = hole[holeNum];
    hole[holeNum] = 0;
    for (;;) {
	if (--holeNum < 0) {
	    if (holeOwner == player) {
		++b->mancala[player];
		if (!--seeds) {
		    nextPlayer = player;
		    break;
		}
	    }
	    holeOwner ^= 1;
	    hole = b->hole[holeOwner];
	    holeNum = MAX_HOLES - 1;
	} 
	++hole[holeNum];
	if (!--seeds) {
	    if (holeOwner == player) {
		oppositeHoleNum = MAX_HOLES - 1 - holeNum;
		oppositeHole = b->hole[player ^ 1];
		if (hole[holeNum] == 1 && oppositeHole[oppositeHoleNum]) {
		    b->mancala[player] += 1 + oppositeHole[oppositeHoleNum];
		    oppositeHole[oppositeHoleNum] = 0;
		    hole[holeNum] = 0;
		}
	    }
	    break;
	}
    }
    return nextPlayer;
}


mancala.h
/* $Id: mancala.h,v 1.2 2007-06-13 18:29:52 sverrehu Exp $ */
#ifndef MANCALA_H
#define MANCALA_H

/*
 *  Mancala board setup. According to tradition, there are 6 small holes
 *  and 4 stones pr hole initially.
 */
#define MAX_HOLES  6
#define STONES_PR_HOLE 4

/*
 *  We use board stack position 0 as the `real' board, so we need a stack
 *  that is one level deeper than the max recursion depth.
 */
#define MAX_DEPTH  20
#define STACK_LEVELS (MAX_DEPTH + 1)

/*
 *  A Move is identified by a hole number.
 */
typedef struct Move {
    int hole;
} Move;

/*
 *  The Board is described by the number of stones in each of the small
 *  holes, and in the mancalas (larger holes). In this setup, hole[x][0]
 *  is the hole closest to the mancala[x] for player x.
 */
typedef char StoneCount;
typedef struct Board {
    StoneCount hole[2][MAX_HOLES];
    StoneCount mancala[2];
} Board;

void       initGame(int stones_pr_hole);
StoneCount getHole(Player player, int hole);
StoneCount getMancala(Player player);
int        legalMove(Player player, PMove move);
int        checkAndFixWin(Player *winner);
PMove      getBestMove(Player player, int maxPly);
Player     doMove(Player player, PMove move);

#endif


rulestxt.c
/* $Id: rulestxt.c,v 1.5 2012-06-18 18:11:52 sverrehu Exp $ */
/**************************************************************************
 *
 *  FILE            rulestxt.c
 *  MODULE OF       The board game Mancala
 *
 *  DESCRIPTION     Compiled-in instructions used by the X-version.
 *
 *  WRITTEN BY      Sverre H. Huseby
 *
 **************************************************************************/

/**************************************************************************
 *                                                                        *
 *                        P U B L I C    D A T A                          *
 *                                                                        *
 **************************************************************************/

char *textRules =
"Note that there are lots of different rule sets for Mancala. Our\n"
"implementation uses this one:\n"
"\n"
"Setup: The board consists of 2 rows of 6 bins and 2 mancalas. Each\n"
"   player is allocated the row of bins closest to him and the mancala\n"
"   to his right. Initially each bin contains 4 stones, and the\n"
"   mancalas are empty.\n"
"\n"
"Object: The object of the game is to place as many stones as possible\n"
"   in one's mancala before the game ends. This happens when a player,\n"
"   at his turn, has no legal moves, meaning that all his holes are\n"
"   empty. The other player then gets to put any remaining stones on\n"
"   his side into his mancala.\n"
"\n"
"Moving: At his turn, the player picks up all the stones in any of his\n"
"   non-empty bins and starts to sow them by placing one in each of the\n"
"   preceding bins, starting with the one on the right and continuing\n"
"   counterclockwise around the board, skipping the oponents mancala\n"
"   (but not his own).\n"
"\n"
"Extra move: If the last stone is sown in one's mancala one gets to\n"
"   move again.\n"
"\n"
"Capturing: If the last stone is sown in an empty bin belonging to the\n"
"   player, he captures all the stones lying in the bin directly across\n"
"   this bin. These stones and the capturing stone are then placed in\n"
"   the mancala. If the bin directly across is empty, nothing special\n"
"   happens.\n"
"\n"
"@_\n"
"\n"
"Comments? Send mail to either of\n"
"   shh@thathost.com\n"
"   glennli@simula.no\n"
"\n"
;



textmain.c
/* $Id: textmain.c,v 1.2 2007-06-13 18:29:52 sverrehu Exp $ */
/**************************************************************************
 *
 *  FILE            textmain.c
 *  MODULE OF       The board game Mancala
 *
 *  DESCRIPTION     Main function / frontend for using a textbased display.
 *
 *  WRITTEN BY      Sverre H. Huseby
 *
 **************************************************************************/

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#include "minimax.h"
#include "mancala.h"

/**************************************************************************
 *                                                                        *
 *                   P R I V A T E    F U N C T I O N S                   *
 *                                                                        *
 **************************************************************************/

/*-------------------------------------------------------------------------
 *
 *  NAME          showBoard
 *
 *  FUNCTION      Display the current board on screen.
 *
 *  DESCRIPTION   Very simple thing.
 */
static void showBoard(void)
{
    int q;

    printf("    ");
    for (q = 0; q < MAX_HOLES; q++)
	printf((q < MAX_HOLES - 1) ? "%c:%2d | " : "%c:%2d\n",
	       'a' + MAX_HOLES - q - 1, getHole(1, q));
    printf(" %2d ", getMancala(1));
    for (q = 0; q < MAX_HOLES; q++)
	printf((q < MAX_HOLES - 1) ? "-----+-" : "----");
    printf(" %2d\n", getMancala(0));
    printf("    ");
    for (q = 0; q < MAX_HOLES; q++)
	printf((q < MAX_HOLES - 1) ? "%c:%2d | " : "%c:%2d\n",
	       'a' + q, getHole(0, MAX_HOLES - q - 1));
}



/**************************************************************************
 *                                                                        *
 *                    P U B L I C    F U N C T I O N S                    *
 *                                                                        *
 **************************************************************************/

int main(int argc, char *argv[])
{
    char   s[80];
    char   playerName[2][20] = { "player at bottom", "player at top" };
    int    maxPly[2] = { MAX_DEPTH, MAX_DEPTH }, status;
    int    stones_pr_hole = STONES_PR_HOLE;
    PMove  move;
    Move   move2;
    Player player, winner;
    
    if (argc == 2)
	maxPly[0] = maxPly[1] = atoi(argv[1]);
    else if (argc == 3) {
	maxPly[1] = atoi(argv[1]);
	maxPly[0] = atoi(argv[2]);
    } else if (argc == 4) {
	maxPly[1] = atoi(argv[1]);
	maxPly[0] = atoi(argv[2]);
	stones_pr_hole = atoi(argv[3]);
    } else {
	printf("usage: mancala max-ply\n"
	       "       mancala max-ply-top max-ply-bottom\n"
	       "       mancala max-ply-top max-ply-bottom number-of-stones\n"
	       "\n"
	       "max-ply of 0 indicates human user.\n"
	       "Player at top starts.\n"
	       "\n");
	return 1;
    }
    if (stones_pr_hole < 1) {
        fprintf(stderr,
                "mancala: the number of stones per hole must be greater "
                "than zero\n");
        return 1;
    }
    if (maxPly[0] < 0 || maxPly[1] < 0) {
        fprintf(stderr,
                "mancala: max-ply must be no less than zero\n");
        return 1;
    }
    initGame(stones_pr_hole);
    player = 1;
    for (;;) {
	showBoard();
	printf("\n");
	if ((status = checkAndFixWin(&winner)) != 0) {
	    printf("=== Game over; ");
	    if (status == 1)
		printf("%s wins", playerName[winner]);
	    else
		printf("we have a draw");
	    printf(" ===\n\n");
	    showBoard();
	    printf("\n");
	    return 0;
	}
	printf("Move for %s: ", playerName[player]);
	if (!maxPly[player]) {
	    move = &move2;
	    for (;;) {
		if (fgets(s, 80, stdin) == NULL) {
                    printf("\nBye.\n");
                    return 0;
                }
		move->hole = 'A' + (MAX_HOLES - 1) - toupper(s[0]);
		if (legalMove(player, move))
		    break;
		printf("not a legal move. try again: ");
	    }
	    printf("\n");
	} else {
	    move = getBestMove(player, maxPly[player]);
	    printf("%c\n\n", 'A' + (MAX_HOLES - 1) - move->hole);
	}
	player = doMove(player, move);
    }
    return 0;
}


xmain.c
/* $Id: xmain.c,v 1.7 2012-06-18 18:07:40 sverrehu Exp $ */
/**************************************************************************
 *
 *  FILE            xmain.c
 *  MODULE OF       The board game Mancala
 *
 *  DESCRIPTION     Main function / frontend for using an X/XForms-based
 *                  display.
 *
 *                  This code is _really_ (I mean: REALLY) dirty...
 *                  The main purpose of this is to make someone play the
 *                  game. For those interested in the gaming (as I am),
 *                  please take a look at the other sourcefiles instead.
 *                  I'm not at all proud of this file, it's my first
 *                  attempt on using the XForms library, and it was done
 *                  in a hurry.
 *
 *                  I don't like event driven programming...
 *
 *  WRITTEN BY      Sverre H. Huseby
 *
 **************************************************************************/

#include <stdlib.h>
#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <forms.h>

#include "minimax.h"
#include "mancala.h"
#include "xform.h"

#define flushForms fl_check_only_forms

extern char *textRules;

/**************************************************************************
 *                                                                        *
 *                       P R I V A T E    D A T A                         *
 *                                                                        *
 **************************************************************************/

static FD_mancala *frm;
static FD_rules   *frmRules;
static FL_OBJECT  *frmMancala[2];
static FL_OBJECT  *frmHole[2][MAX_HOLES];
static FL_OBJECT  *frmLight[2][MAX_HOLES];
static const char *playerName[2] = { "the human player", "the computer" };
static int        maxPly[2] = { 0, 4 };
static int        rulesDisplayed = 0;
static int        stones_pr_hole = STONES_PR_HOLE;

/*
 *  Player side and hole chosen by human player.
 */
static Player movePlayer;
static int    moveHole = -1;

static Player player;
static int    status;


/**************************************************************************
 *                                                                        *
 *                   P R I V A T E    F U N C T I O N S                   *
 *                                                                        *
 **************************************************************************/

/*-------------------------------------------------------------------------
 *
 *  NAME          showBoard
 *
 *  FUNCTION      Display the current board on screen.
 *
 *  DESCRIPTION   Very simple thing.
 */
static void showBoard(void)
{
    Player p;
    int    q;
    char   s[10];

    fl_freeze_form(frm->mancala);
    for (p = 0; p < 2; p++) {
	sprintf(s, "%d", getMancala(p));
	fl_set_object_label(frmMancala[p], s);
	for (q = 0; q < MAX_HOLES; q++) {
	    sprintf(s, "%d", getHole(p, q));
	    fl_set_object_label(frmHole[p][q], s);
	}
    }
    fl_unfreeze_form(frm->mancala);
    flushForms();
}

/*-------------------------------------------------------------------------
 *
 *  NAME          setMessage
 *
 *  FUNCTION      Set the message of the status line.
 *
 *  INPUT         format, ...
 *                        Arguments used as with printf().
 */
static void setMessage(const char *format, ...)
{
    va_list ap;
    char    s[120];

    va_start(ap, format);
    vsprintf(s, format, ap);
    va_end(ap);
    fl_set_object_label(frm->textStatus, s);
    flushForms();
}

/* set hole to -1 to turn all off */
static void setLastMove(Player player, int hole)
{
    static Player   lastPlayer;
    static int      lastHole = -1;
#ifdef WORKS_ON_SGI_NOW
    static int      colorSet = 0;
#endif
    static FL_COLOR origCol1 = FL_COL1, origCol2 = FL_BOTTOM_BCOL;

    /*
     *  For some reason, SGI fucks up with segmenation fault during
     *  the next few lines. Harcoding it instead...
     */
#ifdef WORKS_ON_SGI_NOW
    if (!colorSet) {
	origCol1 = frmLight[player][hole]->col1;
	origCol2 = frmLight[player][hole]->col2;
	colorSet = 1;
    }
#endif
    if (lastHole >= 0)
	fl_set_object_color(frmLight[lastPlayer][lastHole],
			    origCol1, origCol2);
    if (hole >= 0)
	fl_set_object_color(frmLight[player][hole],
			    FL_RED, FL_BOTTOM_BCOL);
    lastPlayer = player;
    lastHole = hole;
    flushForms();
}

static void flashMove(Player player, int hole)
{
    int q;

    setLastMove(0, -1);
    for (q = 5; q > 0; q--) {
	setLastMove(player, hole);
	fl_msleep(100);
	setLastMove(0, -1);
	fl_msleep(100);
    }
}

static Player computerPlayer(void)
{
    if (maxPly[0])
	return 0;
    if (maxPly[1])
	return 1;
    return -1;
}

static void formInit(void)
{
    char s[20];
    
    frm = create_form_mancala();
    sprintf(s, "version %s", VERSION);
    fl_set_object_label(frm->textVer, s);
    frmMancala[0] = frm->mancalaB;
    frmHole[0][0] = frm->holeB0;
    frmHole[0][1] = frm->holeB1;
    frmHole[0][2] = frm->holeB2;
    frmHole[0][3] = frm->holeB3;
    frmHole[0][4] = frm->holeB4;
    frmHole[0][5] = frm->holeB5;
    frmMancala[1] = frm->mancalaT;
    frmHole[1][0] = frm->holeT0;
    frmHole[1][1] = frm->holeT1;
    frmHole[1][2] = frm->holeT2;
    frmHole[1][3] = frm->holeT3;
    frmHole[1][4] = frm->holeT4;
    frmHole[1][5] = frm->holeT5;
    frmLight[0][0] = frm->lightB0;
    frmLight[0][1] = frm->lightB1;
    frmLight[0][2] = frm->lightB2;
    frmLight[0][3] = frm->lightB3;
    frmLight[0][4] = frm->lightB4;
    frmLight[0][5] = frm->lightB5;
    frmLight[1][0] = frm->lightT0;
    frmLight[1][1] = frm->lightT1;
    frmLight[1][2] = frm->lightT2;
    frmLight[1][3] = frm->lightT3;
    frmLight[1][4] = frm->lightT4;
    frmLight[1][5] = frm->lightT5;
    fl_set_slider_bounds(frm->slidLevel, 1.0, 9.0);
    fl_set_slider_step(frm->slidLevel, 1.0);
    fl_set_slider_precision(frm->slidLevel, 0);
    /* HERE: change if not always human vs. computer */
    fl_set_slider_value(frm->slidLevel, maxPly[computerPlayer()]);

    frmRules = create_form_rules();
    fl_set_browser_fontsize(frmRules->textRules, FL_NORMAL_SIZE);
    fl_set_browser_fontstyle(frmRules->textRules, FL_NORMAL_STYLE);
    fl_add_browser_line(frmRules->textRules, textRules);
    
    fl_show_form(frm->mancala, FL_PLACE_SIZE, FL_FULLBORDER, "Mancala");
}

static void formFinish(void)
{
    fl_hide_form(frm->mancala);
    if (rulesDisplayed)
	fl_hide_form(frmRules->rules);
}

static void ourInitGame(Player starter)
{
    initGame(stones_pr_hole);
    player = starter;
    status = 0;
    fl_set_object_label(frm->playerB, playerName[0]);
    fl_set_object_label(frm->playerT, playerName[1]);
    fl_set_object_label(frm->startB, starter == 0 ? "Starter" : "");
    fl_set_object_label(frm->startT, starter == 1 ? "Starter" : "");
    fl_set_object_label(frm->winB, "");
    fl_set_object_label(frm->winT, "");
}

static void formHandler(void)
{
    PMove  move;
    Move   move2;
    Player winner;

    ourInitGame(0);
    for (;;) {
	showBoard();
	if (!status) {
	    if ((status = checkAndFixWin(&winner)) != 0) {
		if (status == 1) {
		    setMessage("Game over; %s wins", playerName[winner]);
		    fl_set_object_label(winner == 0 ? frm->winB : frm->winT,
					"Winner");
		} else
		    setMessage("Game over; we have a draw");
	    } else {
		if (!maxPly[player]) {
		    setMessage("It's %s's turn", playerName[player]);
		    fl_do_forms();
		    move = &move2;
		    if (moveHole >= 0) {
			if (movePlayer == player) {
			    move->hole = moveHole;
			    if (legalMove(player, move)) {
				flashMove(player, move->hole);
				player = doMove(player, move);
			    } else {
				setMessage("Not a legal move. Try again!");
				fl_msleep(1500);
			    }
			} else {
			    setMessage("Stick to your side, please!");
			    fl_msleep(1500);
			}
			moveHole = -1;
		    }
		} else {
		    setMessage("%s is thinking...", playerName[player]);
		    move = getBestMove(player, maxPly[player]);
		    setMessage("");
		    flashMove(player, move->hole);
		    player = doMove(player, move);
		}
	    }
	} else
	    fl_do_forms();
    }
}

/**************************************************************************
 *                                                                        *
 *                    P U B L I C    F U N C T I O N S                    *
 *                                                                        *
 **************************************************************************/

/*------------------------------------------------------------------------*
 |                Callbacks for the objects in the form                   |
 *------------------------------------------------------------------------*/

void doRules(FL_OBJECT *ob, long arg)
{
    if (rulesDisplayed)
	return;
    fl_show_form(frmRules->rules, FL_PLACE_FREE, FL_FULLBORDER,
		 "Mancala Rules");
    fl_deactivate_object(frm->buttRules);
    rulesDisplayed = 1;
}

void doRulesDone(FL_OBJECT *ob, long arg)
{
    if (!rulesDisplayed)
	return;
    fl_hide_form(frmRules->rules);
    fl_activate_object(frm->buttRules);
    rulesDisplayed = 0;
}

void doHole(FL_OBJECT *ob, long arg)
{
    Player player;
    int    hole;

    if (moveHole >= 0) {
	setMessage("Hold your horses! "
		   "I haven't done your previous move yet!");
	fl_msleep(1000);
    } else {
	player = arg / 100;
	hole = arg % 100;
	movePlayer = player;
	moveHole = hole;
    }
    fl_trigger_object(frm->dummyButton);
}

void doLevel(FL_OBJECT *ob, long arg)
{
    double v;
    Player comp;

    if ((comp = computerPlayer()) < 0)
	return;
    v = fl_get_slider_value(ob);
    maxPly[comp] = (int) v;
}

void doNewGame(FL_OBJECT *ob, long arg)
{
    ourInitGame(arg);
    fl_trigger_object(frm->dummyButton);
}

void doQuit(FL_OBJECT *ob, long arg)
{
    formFinish();
    exit(0);
}

/**************************************************************************/
#define stringify(a) #a
#define stringize(a) stringify(a)

int main(int argc, char *argv[])
{
    static FL_CMD_OPT cmdopt [] = {
        { "-stones", "*.stones", XrmoptionSepArg, 0 }};
    static FL_resource res [] = {
        { "stones", "XMancala", FL_INT, &stones_pr_hole,
          stringize(STONES_PR_HOLE), 0 }};
    fl_initialize(&argc, argv, "XMancala",
                  cmdopt, sizeof(cmdopt) / sizeof(FL_CMD_OPT));
    fl_get_app_resources(res, sizeof(res ) / sizeof(FL_resource));
    if (stones_pr_hole < 1) {
        fprintf(stderr,
                "xmancala: the number of stones per hole must be "
                "greater than zero\n");
        return 1;
    }
    formInit();
    formHandler();
    formFinish();
    return 0;
}
